File: backend/app/schemas/anulacion.py
from __future__ import annotations

import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class LineaImpactoRead(BaseModel):
    """Una línea del libro de caja que la anulación va a revertir."""

    model_config = ConfigDict(from_attributes=True)

    fecha: str
    moneda: str
    tipo: str
    categoria: str
    monto: Decimal
    detalle: str | None = None


class ImpactoRead(BaseModel):
    """Qué pasa (o pasaría) al anular una operación.

    El panel pide esto antes de mostrar el diálogo de confirmación: así el
    operador ve exactamente qué movimientos de caja se van a deshacer, en vez de
    apretar "Eliminar" a ciegas.
    """

    model_config = ConfigDict(from_attributes=True)

    entidad: str
    entidad_id: uuid.UUID
    descripcion: str
    puede_anular: bool
    bloqueo: str | None = None
    lineas: list[LineaImpactoRead] = Field(default_factory=list)
    arrastra: list[str] = Field(default_factory=list)


class AnularRequest(BaseModel):
    """Toda anulación queda firmada: quién y por qué."""

    operador_id: str = Field(min_length=1, max_length=80)
    motivo: str = Field(min_length=1)


class RevertirRequest(AnularRequest):
    """Misma firma que anular: revertir también es una operación auditada."""

File: frontend/src/api/anulacion.ts
import { apiFetch } from './client'

/** Tipos de entidad que acepta el motor de anulación (backend `_ENTIDADES`). */
export type EntidadAnulable =
  | 'cheque'
  | 'prestamo'
  | 'movimiento_efectivo'
  | 'fiado'
  | 'deuda_simple'
  | 'pasivo'
  | 'gasto'
  | 'ajuste_caja'

/** Una línea del libro de caja que la anulación va a revertir. */
export interface LineaImpacto {
  fecha: string
  moneda: string
  tipo: 'INGRESO' | 'EGRESO'
  categoria: string
  monto: string
  detalle: string | null
}

/**
 * Qué pasa (o pasaría) al anular. Se pide ANTES de mostrar la confirmación para
 * que el operador vea qué movimientos de caja se deshacen, en vez de apretar
 * "Eliminar" a ciegas.
 */
export interface Impacto {
  entidad: EntidadAnulable
  entidad_id: string
  descripcion: string
  puede_anular: boolean
  bloqueo: string | null
  lineas: LineaImpacto[]
  arrastra: string[]
}

export const previsualizarAnulacion = (
  entidad: EntidadAnulable,
  id: string,
): Promise<Impacto> =>
  apiFetch<Impacto>(`/anulaciones/${entidad}/${encodeURIComponent(id)}`)

export const anular = (
  entidad: EntidadAnulable,
  id: string,
  payload: { operador_id: string; motivo: string },
): Promise<Impacto> =>
  apiFetch<Impacto>(`/anulaciones/${entidad}/${encodeURIComponent(id)}`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

/**
 * Devuelve un cheque terminal a EN_CARTERA. No lo elimina: queda disponible para
 * volver a venderse o fiarse.
 */
export const revertirCheque = (
  id: string,
  payload: { operador_id: string; motivo: string },
): Promise<unknown> =>
  apiFetch(`/cheques/${encodeURIComponent(id)}/revertir`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
